src/main/resources/org/jenkinsci/plugins/durabletask/heartbeat-launcher.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+func checkHeartbeatErr(err error) {
+	checkErr("heartbeat", err)
+}
+
+func checkLauncherErr(err error) {
+	checkErr("launcher", err)
+}
+
+func checkErr(process string, err error) {
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "(%v) %v\n", process, err.Error())
+	}
+}
+
+// In the event the launch script is being intentionally terminated, this signal catcher routine
+// ensures that the main program will stay alive and record the exit code of the script.
+// Without it, there are possible race conditions in which the main program will terminate before
+// the script, and thus not record the exit code to the result file
+func signalCatcher(sigChan chan os.Signal) {
+	for sig := range sigChan {
+		fmt.Printf("(sig catcher) caught: %v\n", sig)
+	}
+}
+
+func launcher(wg *sync.WaitGroup, doneChan chan bool,
+	scriptPath string, logPath string, resultPath string, outputPath string) {
+
+	defer wg.Done()
+	recordExit := fmt.Sprintf("; echo $? > %v.tmp; mv %v.tmp %v; wait", resultPath, resultPath, resultPath)
+	scriptWithExit := scriptPath + recordExit
+	scriptCmd := exec.Command("/bin/sh", "-c", scriptWithExit)
+	logFile, err := os.Create(logPath)
+	checkLauncherErr(err)
+	defer logFile.Close()
+
+	if outputPath != "" {
+		// capturing output
+		outputFile, err := os.Create(outputPath)
+		checkLauncherErr(err)
+		defer outputFile.Close()
+		scriptCmd.Stdout = outputFile
+	} else {
+		scriptCmd.Stdout = logFile
+	}
+	if outputPath != "" {
+		// capturing output
+		scriptCmd.Stderr = logFile
+	} else {
+		// Note: pointing to os.Stdout will not capture all err logs and fail unit tests
+		scriptCmd.Stderr = scriptCmd.Stdout
+	}
+	// Prevents script from being terminated if program gets terminated
+	scriptCmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}
+	// Allows child processes of the script to be killed if kill signal sent to script's process group id
+	scriptCmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
```

Comment:
Good catch, yeah, i'll have to do a separate check on windows when building it. Oddly enough, the script would survive when i killed the main program still, so maybe that's because of setpgid behavior.

Reply:
fyi, apparently setting both pgid and setsid cause a panic